proposals/0033-data-collections.md

```diff
@@ -0,0 +1,243 @@
+
+Thanks to the generic function name, this can be used as a replacement for `getEntryBySlug()` as well. When querying a content collection, `getEntry()` uses `slug` as the identifier:
+
+
+```astro
+---
+// Option 1: separate args
+const welcomePost = await getEntry('blog', 'welcome');
+// Option 2: object keys
+const welcomePost = await getEntry({ collection: 'blog', slug: 'welcome' });
+const Content = await welcomePost.render();
+---
+
+<h1>{welcomePost.data.title}</h1>
+<Content />
+```
+
+### `id` vs `slug`
+
+You may have noticed two competing identifiers depending on the collection type: `id` for data, and `slug` for content. This is an inconsistency we'd like to address in a future RFC, with `id` becoming the new standard for identifying collection entries. For now, `slug` will remain on content collections to make the introduction of data collections non-breaking.
+
+**Full background:** `slug` was originally introduced alongside the content `id` to have a URL-friendly version of the file path, which can be passed to `getStaticPaths()` for route generation. Data collections are not intended to be used as routes, so we don't want to perpetuate this pattern. `slug` also "slugifies" the file path by removing capitalization and replacing spaces with dashes. If we added this processing to data collection IDs, [collection references](https://github.com/withastro/roadmap/blob/d89e2a4c28379108501aa6bf40d2f8d93d81ad02/proposals/0034-collection-references.md) will be less intuitive to use (i.e. "do I include spaces in the referenced ID here?").
+
+## Internals
+
+Data collections should following the API design of content collections with a stripped-down featureset. To wire up data collections, we will introduce an internal utility that mirrors our `addContentEntryType()` integration function. JSON and YAML will be preconfigured by Astro, as a way to dogfood a generic API for wiring up any data collection type in the future. This example shows an internal implementation for `.json` files:
+
+```ts
+addDataEntryType({
+  extensions: ['.json'],
+  getEntryInfo({ contents, fileUrl }) {
+    // Handle empty JSON files, which cause `JSON.parse` to throw
+    if (contents === undefined || contents === '') return { data: {} };
+
+    const data = JSON.parse(contents);
+
+    if (data == null || typeof data !== 'object')
+      throw new Error(`JSON collection entry ${fileUrl.pathname} must be an object.`);
+
+    return { data };
+  },
+});
+```
+
+Then, we will update our type generator to recognize these data-specific file extensions. This should also raise errors when collections are misconfigured (i.e. `type: 'data'` is missing from the config file) and when a mix of content and data in the same collection is detected.
```

Comment:
> This should also raise errors ... a mix of content and data in the same collection is detected.

I could see this being a bit of a blocker for the easy upgrade path of a unified `src/content` directory

If I have a large data collection and only need to upgrade one to use body content for some specific use case, I'd end up having to update every item in the collection to `.md` files

I can't think of a *great* example of this at the moment so it may very well be a corner case, feel free to ignore if it isn't a use case worth covering!

Reply:
Fair enough! There may be a case to allow content inside _data_ collections, since data collections have a subset of content collections' properties. Though it sounds like you're proposing the other way. In that case, properties like `render()` and `body` would be added to every collection item and stubbed out with `undefined`'s, which I worry about for usability. Also not sure about "mixing" return types based on the entry.

My thinking is this:
- Move forward with collection-level types for launch
- Consider allowing content collections in data collections as a fast follow
- Look into the "singleton" discussion here (https://github.com/withastro/roadmap/discussions/449) for one-off `.md` files as you're describing here. I think there's a use case to allowing one-off entries with a shared schema, though it's something we could tackle separately.

Let me know if that makes sense, or if I'm missing anythign!

Reply:
Yep, another case of "defensively make restrictive, open up if people want it" 👍